### PR TITLE
Controller to check the different endpoints for authentication

### PR DESCRIPTION
**Pull request by vareti**
This PR provides a controller that regularly checks health of the oauth-server using its route, service IP and pod IPs

### CODE REVIEW

File: pkg/controllers/endpointaccessible/endpoint_accessible_controller.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	serviceInformer := corev1Informers.Services().Informer()
+
+	endpointsLister := corev1Informers.Endpoints().Lister()
+	endpointsInformer := corev1Informers.Endpoints().Informer()
+
+	// collects endpoints through which oauth-server can be reached
+	endpointListFunc := func() ([]string, error) {
+		var results []string
+
+		routes, err := listOAuthServerRoutes(routeLister, recorder)
```
Comment (deads2k) [inline pkg/controllers/endpointaccessible/endpoint_accessible_controller.go:46]: one controller for each of these categories please.  That way we can get different degraded reasons.

**Comment by deads2k** [comment]
The conditions exist. In combination with the resourcewatch this will diagnose our issues.

/approve
/assign @stlaz

File: pkg/controllers/endpointaccessible/endpoint_accessible_controller.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+type endpointAccessibleController struct {
+	operatorClient v1helpers.OperatorClient
+	endpointListFn EndpointListFunc
+}
+
+type EndpointListFunc func() ([]string, error)
+
+// NewEndpointAccessibleController returns a controller that checks if the endpoints
+// listed by endpointListFn are reachable
+func NewEndpointAccessibleController(
```
Comment (deads2k) [inline pkg/controllers/endpointaccessible/endpoint_accessible_controller.go:141]: in a followup PR, I'd like to see this moved to a separate package.  It's the reusable collateral.

Comment (vareti) [reply pkg/controllers/endpointaccessible/endpoint_accessible_controller.go:141]: I made a note of it and will move it to separate package in follow up PR.

**Comment by vareti** [comment]
/hold

look genuine failures. Need to bump openshift/api to fix it.

**Comment by vareti** [comment]
/hold cancel

https://github.com/openshift/kubernetes/pull/263 changed the validation rules for oauth cluster config. This resulted in failing e2e-operator tests. Bumping the openshift/api repo fixed it.

**Pull request merged**